emul: Add EMUL_DT_GET

alevkoy opened the pull request:
Implement EMUL_DT_GET, analogous to DEVICE_DT_GET. Add a test.

Signed-off-by: Abe Levkoy <alevkoy@google.com>

alevkoy commented:
I tried to do this analogously to `DEVICE_DT_GET` and its test.

mbolivar-nordic reviewed (dismissed):
I haven't followed the development of the emul.h API but I have no objections

alevkoy commented:
Re: test failures above:
* Locally for me, `tests/net/ipv6/net.ipv6` and `tests/net/socket/tcp/net.socket.tcp` pass on `mps2_an385`. Provisionally, I'm assuming these are flakes.
* The vast majority of the above failures are CMake failures affecting my new test, of the form
  ```
  CMake Error at ../../../../cmake/modules/extensions.cmake:3882 (message):
  zephyr_linker_section_configure(SECTION ...) given unknown arguments:
  ALIGN_WITH_INPUT
  Call Stack (most recent call first):
  ../../../../cmake/linker_script/common/common-rom.cmake:147 (zephyr_linker_section_configure)
  ../../../../cmake/linker_script/arm/linker.cmake:99 (include)
  ../../../../CMakeLists.txt:118 (include)


  -- Configuring incomplete, errors occurred!
  ```
  I can reproduce these locally on various platforms.

I modeled my test on `tests/lib/devicetree/devices`, and it doesn't seem to have this problem. I will investigate further.

As an aside, I might have expected my
```
integration_platforms:
  - native_posix
```
snippet in testcase.yaml to restrict Twister to running on `native_posix`, which passes, but I see that the devices test does not work this way either.

yperess reviewed (changes requested):
I think this is only working because your call to EMUL_DEFINE is in the same file as EMUL_DT_GET. Please try this with an existing emulator. A good test to add this to would be in tests/driver/sensor/accel

yperess commented:
> I think this is only working because your call to EMUL_DEFINE is in the same file as EMUL_DT_GET. Please try this with an existing emulator. A good test to add this to would be in tests/driver/sensor/accel

https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/sensor/accel/src/main.c

alevkoy commented:
> I think this is only working because your call to EMUL_DEFINE is in the same file as EMUL_DT_GET. Please try this with an existing emulator. A good test to add this to would be in tests/driver/sensor/accel

You're right. I'll see what else I can come up with.

alevkoy committed:
emul: Add EMUL_DT_GET

Implement EMUL_DT_GET, analogous to DEVICE_DT_GET. Add a test.

Signed-off-by: Abe Levkoy <alevkoy@google.com>

alevkoy committed:
emul: Expose struct emul objects

Give struct emul objects defined by EMUL_DEFINE external linkage.
Generate a struct emul prototype for each node in the device tree in
emul.h.

Define EMUL_DT_NAME_GET and deprecate EMUL_REG_NAME, which was only used
in EMUL_DEFINE.

This partially addresses #45976 by making it possible to statically
refer to struct emuls.

Signed-off-by: Abe Levkoy <alevkoy@google.com>

alevkoy commented:
I believe what I just pushed addresses Yuval's concern. I'm still getting the previously mentioned `ALIGN_WITH_INPUT` error on some platforms, which I will now investigate.

alevkoy committed:
cmake: Fix typo in common-rom.cmake

ALIGN_WITH_INPUT is an argument for zephyr_linker_section, not
zephyr_linker_section_configure. Fix the calls for section .emulators
accordingly.

Signed-off-by: Abe Levkoy <alevkoy@google.com>

alevkoy commented:
I think I fixed the `ALIGN_WITH_INPUT` errors previously seen. Torsten, see my first commit.

yperess commented on tests/lib/devicetree/emul/src/emulator.c:
```diff
@@ -0,0 +1,19 @@
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+
+static int emul_init(const struct emul *emul, const struct device *parent)
```
This is the hardest coding guideline to understand so just helping out, you can't use `emul` as both the type and the name of the variable. Suggest: `dev_emul` for the variable name. This comes from C++ where the `struct` key word is optional when declaring a variable, so you could write `emul emul` which is confusing.

yperess commented on tests/lib/devicetree/emul/src/emulator.c:
```diff
@@ -0,0 +1,19 @@
+ * This emulator definition is deliberately in a different compilation unit than
+ * test_emuL_dt_get to verify that emulators are externally accessible.
+ */
+EMUL_DEFINE(emul_init, DT_NODELABEL(test_gpio_0), NULL, NULL);
```
Can we include an `EMUL_DT_DEFINE` to mirror `DEVICE_DT_DEFINE`. The contract of `EMUL_DT_GET` should be that it will map 1:1 with `EMUL_DT_DEFINE`, if the developer used anything else it's on them to make sure that the IDs are aligned. The appropriate test should include:

1. A new binding with a test `compatible` string
2. The test `emulator.c` defining `DT_DRV_COMPAT` to match the compatible string
3. The test `emulator.c` calling `DT_INST_FOREACH_STATUS_OKAY` to initiate one emulator per instance of `DT_DRV_COMPAT`
4. The test `main.c` should find the instance using any of the node identifier get macros (`DT_NODELABEL` is one of them so that works like you have it).

yperess commented on tests/lib/devicetree/emul/src/main.c:
```diff
@@ -0,0 +1,27 @@
+ZTEST(emul, test_emul_dt_get)
+{
+	const struct emul *emul = emul_get_binding(DT_LABEL(TEST_GPIO));
+	const struct emul *emul_static = EMUL_DT_GET(TEST_GPIO);
```
Can you make this `static` so we can verify that the macro resolved as expected?

yperess commented on tests/lib/devicetree/emul/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+    # will mostly likely be the fastest.
+    integration_platforms:
+      - native_posix
+    platform_exclude: hsdk hsdk_2cores thingy52_nrf52832
```
Can you add a comment why these are excluded?

alevkoy committed:
emul: Define EMUL_DT_DEFINE, EMUL_DT_INST_DEFINE

Define these macros similarly to their DEVICE_*_DEFINE counterparts.

Part of #45976.

Signed-off-by: Abe Levkoy <alevkoy@google.com>

alevkoy committed:
emul: Update EMUL_DEFINE usages in emulators

Use EMUL_DT_INST_DEFINE instead.

Part of #45976.

Signed-off-by: Abe Levkoy <alevkoy@google.com>

alevkoy commented:
To avoid multiple overlapping PRs while waiting for the merge window to open, I have added a few commits here pursuant to #45976. @yperess PTAL.

aaronemassey commented on tests/lib/devicetree/emul/src/emulator.c:
```diff
@@ -0,0 +1,22 @@
+
+/*
+ * This emulator definition is deliberately in a different compilation unit than
+ * test_emuL_dt_get to verify that emulators are externally accessible.
```
nit:
test_emuL_dt_get -> test_emul_dt_get

alevkoy replied:
Done.

aaronemassey commented on tests/lib/devicetree/emul/src/emulator.c:
```diff
@@ -0,0 +1,22 @@
+#include <zephyr/drivers/emul.h>
+
+static int emul_init(const struct emul *dev_emul, const struct device *parent)
+{
```
nit:

Add:

```
	ARG_UNUSED(dev_emul);
	ARG_UNUSED(parent);
```

aaronemassey reviewed (dismissed):
Couple nits. Lgtm.

yperess commented on tests/lib/devicetree/emul/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
nit: add `Copyright 2022 Google LLC`

yperess commented on tests/lib/devicetree/emul/app.overlay:
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2022 The Chromium OS Authors
```
nit: upstream we're using `Copyright 2022 Google LLC` AFAIK

yperess commented on tests/lib/devicetree/emul/src/main.c:
```diff
@@ -0,0 +1,30 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device_extern.h>
```
Do we need to include this? I thought this is automatically included in every file by the build system.

yperess reviewed (changes requested):
LGTM, please update the nits by Aaron and myself and I think this is ready.

alevkoy commented:
Added a copyright notice to prj.conf.

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -163,6 +163,7 @@ def write_device_extern_header(device_header_out, edt):
         print('extern "C" {', file=dev_header_file)
         print("#endif", file=dev_header_file)
         print("", file=dev_header_file)
+        print("#include <zephyr/device.h>", file=dev_header_file)
```
Taking a second look at this PR now that the merge window is open. Why is this needed?

You should never need to include device_extern.h directly; it's just an internal sub-header of device.h.

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -164,10 +164,14 @@ def write_device_extern_header(device_header_out, edt):
         for node in sorted(edt.nodes, key=lambda node: node.dep_ordinal):
             print(f"extern const struct device DEVICE_DT_NAME_GET(DT_{node.z_path_id}); /* dts_ord_{node.dep_ordinal} */",
                   file=dev_header_file)
+            print(f"extern const struct emul EMUL_DT_NAME_GET(DT_{node.z_path_id});" \
+                    f" /* dts_ord_{node.dep_ordinal} */",
+                    file=dev_header_file)
```
Like I may have mentioned, I haven't been following the emulator device development carefully, but this looks problematic to me from a layering perspective. device_extern.h is for devices, so this mixing feels off.

Is this a good place for me to start reading? https://docs.zephyrproject.org/latest/hardware/emulator/index.html

alevkoy replied:
I have updated the commit to put the `struct emul`s in their own header file and add the associated boilerplate to `gen_defines.py` and `dts.cmake`.

I thought about trying to combine `write_device_extern_header` and `write_emul_extern_header` into one parameterized function, but constructing the include guard, the name of the public header file, and the actual declaration in a parameterized way seemed error-prone and less readable.

yperess replied:
@alevkoy can this be marked resolved now?

mbolivar-nordic reviewed (changes requested):
Some of this looks a bit fishy; I'd like to put a -1 on this for now until I have a chance to understand this better.

yperess commented on include/zephyr/drivers/emul.h:
```diff
@@ -124,4 +134,6 @@ const struct emul *emul_get_binding(const char *name);
  * @}
  */
 
+#include <emul_extern.h>
```
Is it possible to `#include <generated/emul_extern.h>` for clarity? Do you know if we have just `include/` on the path?

alevkoy replied:
That doesn't work for me right now, so I guess we don't have just `include/` on the include path. Sounds like an opportunity for a future build system PR.

yperess replied:
Thanks for checking, marking this as resolved

yperess reviewed (dismissed):
LGTM, one question about the way we include generated files, but not a blocker since it's the established pattern.

alevkoy commented:
@mbolivar-nordic This is ready for re-review. I think I've addressed your comments.

andyross commented on include/zephyr/drivers/emul.h:
```diff
@@ -67,30 +69,78 @@ struct emul {
+ * @deprecated Use EMUL_DT_DEFINE or EMUL_DT_INST_DEFINE.
+ */
+#define EMUL_DEFINE(init_ptr, node_id, cfg_ptr, data_ptr)                                          \
+	EMUL_DT_DEFINE(node_id, init_ptr, data_ptr, cfg_ptr)
```
Heh, `#define X(a, b, c, d) Y(b, a, d, c)` seems maybe a little cruel for users of the earlier API, but then I guess it's being deprecated and the new argument order looks more conventional.

andyross reviewed (dismissed):
I'm likewise a little out of my depth on this subsystem, but nothing here seems like a red flag.

yperess commented:
@mbolivar-nordic would you like to take another look? It seems that Abe took care of the concerns and move the emulator components to their own `emul_external.h` file

yperess commented:
@mbolivar-nordic any final thoughts? There seem to be overlapping PRs with this one, it would be great to get this in so we can keep the work flowing.

gmarull commented on scripts/dts/gen_defines.py:
```diff
@@ -177,6 +178,33 @@ def write_device_extern_header(device_header_out, edt):
         print("#endif /* DEVICE_EXTERN_GEN_H */", file=dev_header_file)
 
 
+def write_emul_extern_header(emul_header_out, edt):
+    # Generate header that will extern devicetree struct emuls
+
+    with open(emul_header_out, "w", encoding="utf-8") as emul_header_file:
+        print("#ifndef EMUL_EXTERN_GEN_H", file=emul_header_file)
+        print("#define EMUL_EXTERN_GEN_H", file=emul_header_file)
+        print("", file=emul_header_file)
+        print("#ifdef __cplusplus", file=emul_header_file)
+        print('extern "C" {', file=emul_header_file)
+        print("#endif", file=emul_header_file)
+        print("", file=emul_header_file)
+        print("/* Do not include directly. #include <zephyr/drivers/emul.h> instead. */",
+                file=emul_header_file)
+
+        for node in sorted(edt.nodes, key=lambda node: node.dep_ordinal):
+            print(f"extern const struct emul EMUL_DT_NAME_GET(DT_{node.z_path_id});" \
+                    f" /* dts_ord_{node.dep_ordinal} */",
+                    file=emul_header_file)
+
+        print("", file=emul_header_file)
+        print("#ifdef __cplusplus", file=emul_header_file)
+        print("}", file=emul_header_file)
+        print("#endif", file=emul_header_file)
+        print("", file=emul_header_file)
+        print("#endif /* EMUL_EXTERN_GEN_H */", file=emul_header_file)
```
I don't think we should modify device infrastructure because of a subsystem design. If the emulator subsystem decided to be independent of devices (I still don't get the reason device model couldn't be re-used), it should be in all aspects, including any necessary tooling.

alevkoy replied:
If by "device infrastructure," you mean the build processes that operate on devicetrees, I don't think this modifies them significantly. This creates an internal-only header based on the devicetree contents, but we were already doing that for `struct device`. (A previous version of this PR just added to the already-existing header.) The `struct emul` declarations exposed in the header were not public previously, but right now, every `struct emul` definition is based on the devicetree. Basing the corresponding declarations on the devicetree seems like the natural way to go.

If we wanted to define `struct emul`s independently of the devicetree (as we can for `struct device` with `DEVICE_DEFINE`), and we wanted to make those declarations public, we would need to do it via some other mechanism. But to disassociate `struct emul`s from the devicetree entirely, we would need to remove/replace the existing mechanism for defining emulators, because it's based on the devicetree right now.

I think that emulators should probably continue to be based on the devicetree going forward. In any case, I think separating `struct emul` from devicetree is out of scope for this PR.

yperess replied:
@gmarull any more thoughts about this? The tool here is under the dts umbrella and `gen_defines.py` seems to fit since we're mirroring how devices work.

Emulators cannot use devices because they use the same DTS node as the device. A short example:

```
&i2c0 {
  my_dev: my_dev@80 {
    compatible = "test, my_dev";
    reg = <0x80>;
   ...
  };
};
```

This the above, we would create a `struct device` from that node in the driver logic. The emulator (when using the same DTS file) will need to create a parallel emulator that uses the same properties. If we were to use DEVICE_DT_DEFINE we would be creating a second `struct device` with the exact same name. Since both have to live in the same binary, we're using a `struct emul` instead but mirroring how devices work. Now, when the read device driver makes an i2c call, the emulated instance for `i2c0` above will know to route the call to the `struct emul` created for the same node.

If we were to split all the tooling for emulators out, we would be replicating a lot of the same code, since emulators are so tightly coupled to devices in DTS, I do not see a need for them to be done separately. Especially because any large changes to device tooling will also require changes to emulator tooling. It's just adding more liability and making it harder to maintain.

gmarull replied:
Emulators are now independent of devices, it is a subsystem that can be disabled (not like devices for example). So from a design perspective, any of its requirements should also be independent, including scripts. I can't see any relevant function re-use in the current script really, so making it independent should be easy by keeping the same imports in a separate script. I also see that it now runs unconditionally, which is wrong. Why do I need to spend time generating `emul_extern.h` when I don't use emulators?

Keeping things independent eases maintenance as well. If I want to work on a device's script, I should not have to learn about emulators. A similar case happened in the past with PM. It was all mixed with devices, so anyone working on device.h had to learn about PM to make any serious change and not break things. Now we just keep a pointer and the rest is managed in the subsystem.

galak replied:
I've posted https://github.com/zephyrproject-rtos/zephyr/pull/47798 as a way we can make this more generic and only have the EMUL externs come in when EMUL is being used.

alevkoy replied:
#47798 seems like it should work. Further comment there.

alevkoy replied:
@galak Should I be waiting for you to submit #47798, or did you want me to incorporate something like that into this PR?

gmarull commented on include/zephyr/drivers/emul.h:
```diff
@@ -81,27 +82,47 @@ extern const struct emul __emul_list_end[];
+ *
+ * Prefer EMUL_DT_DEFINE
+ */
+#define EMUL_DEFINE(init_ptr, node_id, cfg_ptr, data_ptr)                                          \
```
why not deprecate?

alevkoy replied:
I do in the last commit in this PR.

galak commented on include/zephyr/drivers/emul.h:
```diff
@@ -67,30 +69,78 @@ struct emul {
+ * @deprecated Use EMUL_DT_DEFINE or EMUL_DT_INST_DEFINE.
+ */
+#define EMUL_DEFINE(init_ptr, node_id, cfg_ptr, data_ptr)                                          \
+	EMUL_DT_DEFINE(node_id, init_ptr, data_ptr, cfg_ptr)
```
This should be done the other way around. ` EMUL_DT_DEFINE` should call `EMUL_DEFINE`.

alevkoy replied:
I changed my later commit to remove this instead of deprecating, but I retained the commit structure to keep the commits small.

galak reviewed (changes requested):
Some places needed cleaning up.  Added comment about another way we can handle the `externs`

galak reviewed (commented):
Test probably is better in `tests/subsys/emul`.

alevkoy commented:
> Test probably is better in `tests/subsys/emul`.

Done.

mbolivar-nordic reviewed (dismissed):
@alevkoy I am happy with this. I think we can migrate over to @galak 's way of doing things if we need to. I think the performance overhead of  generating the extra header file is probably negligible compared to the benefit people will get from this functionality, so I'm open to getting it merged if you can commit to continuing to optimize over time.

... and while we're on that subject ...


I don't see an entry for the emulator API in https://docs.zephyrproject.org/latest/develop/api/overview.html. This is a problem; can you send a follow-up PR fixing it?

alevkoy commented:
> @alevkoy I am happy with this. I think we can migrate over to @galak 's way of doing things if we need to. I think the performance overhead of generating the extra header file is probably negligible compared to the benefit people will get from this functionality, so I'm open to getting it merged if you can commit to continuing to optimize over time.

@galak @gmarull I am willing to convert the `struct emul` generation to use Kumar's method once that merges, and I believe this will decouple `gen_defines.py` from `struct emul`. Is that acceptable?

> I don't see an entry for the emulator API in https://docs.zephyrproject.org/latest/develop/api/overview.html. This is a problem; can you send a follow-up PR fixing it?

Added it to my to do list.

mbolivar-nordic commented:
> I am willing to convert the struct emul generation to use Kumar's method once that merges, and I believe this will decouple gen_defines.py from struct emul.

After reading through your comments on Kumar's method, I think it would be better to base future emulator efforts on this instead, since you seem OK with waiting for a bit more infrastructure to get built up underneath you before getting this in: https://github.com/zephyrproject-rtos/zephyr/pull/48092

alevkoy committed:
emul.h: Fix includes

Include devicetree.h, whose contents are already used in this file. Sort
included headers.

Signed-off-by: Abe Levkoy <alevkoy@google.com>

alevkoy commented:
I rebased this temporarily on top of the unmerged commits from #48092, so reviewers can see how this feature will look using the new devicetree API from that PR. Also, and primarily, I want to make it easier for us to pull this feature for a dependent downstream activity before it merges.

I also fixed a few typos I noticed in my earlier commits.

galak commented on include/zephyr/drivers/emul.h:
```diff
@@ -89,7 +90,16 @@ struct emul {
+#define EMUL_DT_NAME_GET(node_id) (_CONCAT(__emulreg_, node_id))
+
+/**
+ * @brief Use the devicetree node identifier as a unique name.
```
not sure there's any need to deprecate this as the emulator API isn't stable.  So we can just remove `EMUL_REG_NAME`.

galak commented on include/zephyr/drivers/emul.h:
```diff
@@ -135,6 +144,49 @@ extern const struct emul __emul_list_end[];
+ *
+ * Prefer EMUL_DT_DEFINE
+ */
+#define EMUL_DEFINE(init_ptr, node_id, cfg_ptr, data_ptr, bus_api)                                 \
```
I think we should just remove `EMUL_DEFINE`.

alevkoy replied:
Added a commit to remove the old names. (I think I had removed this one before, and it snuck back in during a rebase.)

galak reviewed (commented):
should look to remove `emul_get_binding` and just use `EMUL_DT_GET`

alevkoy committed:
emul.h: Remove EMUL_DEFINE, EMUL_REG_NAME

These macros now have more conventionally named replacements, and the
old names are unused.

Part of #45976.

Signed-off-by: Abe Levkoy <alevkoy@google.com>

alevkoy commented:
> should look to remove emul_get_binding and just use EMUL_DT_GET

I've heard enough talk of using `emul_get_binding` that I suspect there's substantial downstream usage, and we'd want some kind of deprecation process. Can that be the subject of a future PR?

gmarull commented on tests/subsys/emul/prj.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_ZTEST_NEW_API=y
+CONFIG_I2C=y
+CONFIG_SENSOR=y
+CONFIG_BMI160=y
```
this is now redundant

alevkoy replied:
Removed redundant entries. Removing any of the remaining ones will cause the build to fail.

gmarull commented on include/zephyr/drivers/emul.h:
```diff
@@ -149,6 +148,27 @@ extern const struct emul __emul_list_end[];
+	EMUL_DT_DEFINE(node_id, init_ptr, data_ptr, cfg_ptr, bus_api)
+
+/**
+ * @def EMUL_DT_INST_DEFINE
+ *
```
not needed

alevkoy replied:
Done.

gmarull commented on include/zephyr/drivers/emul.h:
```diff
@@ -149,6 +149,24 @@ extern const struct emul __emul_list_end[];
 DT_FOREACH_STATUS_OKAY_NODE(Z_MAYBE_EMUL_DECLARE_INTERNAL)
 
+/**
+ * @def EMUL_DT_GET
+ *
```
not needed

alevkoy replied:
Done.

galak commented on tests/subsys/emul/app.overlay:
```diff
@@ -0,0 +1,20 @@
+		compatible = "bosch,bmi160";
+		reg = <0x68>;
+		status = "okay";
+		label = "test-bmi";
```
remove label

alevkoy replied:
I need `label = ` as long as I'm using `emul_get_binding` to validate `EMUL_DT_GET`.

alevkoy replied:
OK, you were right. I just needed to get the name differently for `emul_get_binding`.

alevkoy commented:
Fixed up comment formatting to avoid Doxygen error.

alevkoy commented:
AFAIK, all the test failures are the same as those seen on #48092, and they should be resolved when that merges, and I rebase on top of it.

alevkoy commented:
Rebased now that #48092 has merged.

alevkoy commented:
@galak @gmarull @mbolivar-nordic I think this is ready for review again.

gmarull commented on tests/subsys/emul/README:
```diff
@@ -0,0 +1 @@
+Test cases for the emulator API.
```
This file has no value, can be deleted.

gmarull commented on tests/subsys/emul/prj.conf:
```diff
@@ -0,0 +1,10 @@
+
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
+CONFIG_I2C=y
+CONFIG_SENSOR=y
+CONFIG_BMI160=y
```
this should all be reduced to `CONFIG_SENSOR=y`, see 

https://github.com/zephyrproject-rtos/zephyr/blob/352fde1271c568fa24cb48b4aed20f97b25945b1/drivers/sensor/bmi160/Kconfig#L6-L11

alevkoy replied:
Done.

gmarull commented on tests/subsys/emul/testcase.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  subsys.emul:
+    tags: devicetree
+    # We only need this to run on one platform so use native_posix as it
+    # will mostly likely be the fastest.
```
If I'm not wrong, native_posix is the only supported platform when using emulators (because you really need the I2C emulator), so this comment is misleading.

aaronemassey replied:
> If I'm not wrong, native_posix is the only supported platform when using emulators (because you really need the I2C emulator), so this comment is misleading.

Actually emulators can be used with Qemu as well. I've got a PR showing this that adds a new sbs_gauge emulator.

gmarull replied:
so then platform_allow should reflect this capability as well. You can choose native_posix as an integration_platforms.

alevkoy replied:
My intent here was to avoid redundant testing. Since the feature under test doesn't actually do anything at runtime and only relies on the compiler, linker, and gen_defines.py, I figured testing on one platform would be as good as testing on all of them. But if you want this to run on anything that it can run on, I'm OK with that.

I'm still trying to figure out the best way to do that. My `app.overlay` references `i2c0`, which is a node in `native_posix.dts`, but I found ~400 .dts files in which it is absent. Consequently, I get devicetree errors during configuration for several of the qemu platforms if I remove this `platform_allow`. How is this typically handled?

gmarull replied:
Relying on a hardcoded node isn't a good idea for tests/samples that are meant to run on multiple platforms. A couple of things you could do:

- Filter based on dt, e.g. `filter: dt_compat_enabled("zephyr,i2c-emul-controller")`
- Use an alias for the I2C emulator, e.g. `i2c-emul = &i2c0;`.

gmarull replied:
If it's really a test that you want as a subsys kind of unit test, then it's ok to restrict to native_posix, but then comment needs to be deleted or adjusted to avoid confusions.

yperess replied:
This test should use platform_allow: native_posix

alevkoy replied:
Per OOB discussion with @yperess, I added `depends_on: i2c` and `filter: CONFIG_EMUL`. Currently, this limits the test to `native_posix`. If we support emulated I2C on other platforms in the future, it will run on those too.

alevkoy replied:
Addendum: Apparently, CI runs more platforms than my local twister runs, including some that support I2C and emulators but have different I2C layouts than `native_posix`. For instance, `swan_r5` numbers its busses starting at 1 instead of 0. Coming up with usable devicetree overlays for each such board seems like it would just make this test pointlessly voluminous. I'm OK with `platform_allow: native_posix` if you are.

galak commented:
> > 
> 
> I've heard enough talk of using `emul_get_binding` that I suspect there's substantial downstream usage, and we'd want some kind of deprecation process. Can that be the subject of a future PR?

What does that mean?  I'm ok if you want to mark it deprecated, so we should NOT use it as part of the testcase that this PR adds.

alevkoy commented:
> > I've heard enough talk of using `emul_get_binding` that I suspect there's substantial downstream usage, and we'd want some kind of deprecation process. Can that be the subject of a future PR?
> 
> What does that mean? I'm ok if you want to mark it deprecated, so we should NOT use it as part of the testcase that this PR adds.

I just meant that I didn't want to delete `emul_get_binding` as part of this PR. If you want me to add a deprecated label to `emul_get_binding`, I'll be happy to do that in another PR. IMO, this one has gotten big enough.

I removed `emul_get_binding` from the test and now verify that `EMUL_GET_DT` returned the right emulator by checking the device name. Does that work for you?

galak commented:
> > 
> 
> I just meant that I didn't want to delete `emul_get_binding` as part of this PR. If you want me to add a deprecated label to `emul_get_binding`, I'll be happy to do that in another PR. IMO, this one has gotten big enough.
> 
> I removed `emul_get_binding` from the test and now verify that `EMUL_GET_DT` returned the right emulator by checking the device name. Does that work for you?

Ok, lets follow this up with a PR that deprecates `emul_get_binding` and adds `emul_get_device` that provides a runtime search based on the `struct device` instead of string/label name.

cfriedt commented:
I know that someone has probably made a good argument for this, but why are we using identical DT compat identifiers for both physical and emulated peripherals.

`#define DT_DRV_COMPAT bosch_bmi160`

I would imagine that it's considerably easier to manage things with a separate DT compat for what is (generally) an entirely different piece of code. I.e.

`#define DT_DRV_COMPAT bosch_bmi160_emul`

Sure the device tree changes slightly but differences in device trees are usually solved by way of chosen nodes.

alevkoy commented:
> Ok, lets follow this up with a PR that deprecates emul_get_binding and adds emul_get_device that provides a runtime search based on the struct device instead of string/label name.

Created #48888. Can you elaborate there what you have in mind for `emul_get_device`?

alevkoy commented:
> I know that someone has probably made a good argument for this, but why are we using identical DT compat identifiers for both physical and emulated peripherals. I've always been a bit unsure about that.

I discuss this in some detail in https://github.com/zephyrproject-rtos/zephyr/pull/47388#discussion_r916334794.

cfriedt commented:
> > I know that someone has probably made a good argument for this, but why are we using identical DT compat identifiers for both physical and emulated peripherals. I've always been a bit unsure about that.
> 
> I discuss this in some detail in https://github.com/zephyrproject-rtos/zephyr/pull/47388#discussion_r916334794.
> 

@alevkoy 

Yeah, to be honest, I still think that they should just have different compat strings (e.g. suffixed with `-emul`) and the regular device machinery should be used. It adds an extra layer of unnecessary complexity.

Even if the node does not refer to a physical device, it's still fine to use regular DT bindings. We already have several pseudo devices that use DT bindings.

Differentiating between a physical and emulated peripheral is also not really necessary, since arbitrary device tree overlays can be specified at build time, or even a completely different board can be defined, but differentiating via compat string (or even an `emulator` boolean DT attribute) would probably suffice for the simplest cases.

Thanks for providing an explanation, in any case.

I do very much agree that emulated peripherals, and the ability to control behaviour and configure per-instance behaviour with DT-bindings of those emulated peripherals, is incredibly useful. Just am not super convinced that the extra EMUL macro-machinery is needed.

cc @gmarull @galak @yperess

galak commented:
Marked DNM since we reverted some code this depends on - https://github.com/zephyrproject-rtos/zephyr/pull/48858

alevkoy commented:
Looks like the reland has landed in #48987.

yperess commented:
> > > I know that someone has probably made a good argument for this, but why are we using identical DT compat identifiers for both physical and emulated peripherals. I've always been a bit unsure about that.
> > 
> > I discuss this in some detail in https://github.com/zephyrproject-rtos/zephyr/pull/47388#discussion_r916334794.
> > 
> 
> @alevkoy 
> 
> Yeah, to be honest, I still think that they should just have different compat strings (e.g. suffixed with `-emul`) and the regular device machinery should be used. It adds an extra layer of unnecessary complexity.
> 
> Even if the node does not refer to a physical device, it's still fine. We already have several pseudo devices that use DT bindings, for example, another RFC proposes using DT for network configuration.
> 
> Differentiating between a physical and emulated peripheral is also not really necessary, since arbitrary device tree overlays can be specified at build time via the command line, or even a completely different board can be defined. Differentiating via compat string (or even an `emulator` boolean DT attribute) would probably suffice for the simplest cases.
> 
> Thanks for providing an explanation, in any case.
> 
> I do very much agree that emulated peripherals, and the ability to control and configure per-instance behaviour with DT-bindings of those emulated peripherals, is incredibly useful. Just am not super convinced that the extra EMUL macro-machinery is needed.
> 
> cc @gmarull @galak @yperess 

The reason for the same compat string is to prevent duplication and reduce the chance for errors. As a very simplified example, let's assume we have a device tree like:

```
&i2c0 {
  dev@0x80 {
    reg = <0x80>;
    ...
  };
};
```

In this example, we're able to:
1. Create the device struct/driver
2. Create the emulator
3. Automatically bind the emulator to the bus

If we used a different compat string, I'm assuming you would want a separate node. In that case, you would need to duplicate the properties of the driver node. Also, if you have 2 nodes you can't put them both on the same bus since you can only have 1 node on the same address.

cfriedt commented:
> > > > I know that someone has probably made a good argument for this, but why are we using identical DT compat identifiers for both physical and emulated peripherals. I've always been a bit unsure about that.
> > cc @gmarull @galak @yperess

> The reason for the same compat string is to prevent duplication and reduce the chance for errors. As a very simplified example, let's assume we have a device tree like:

Which errors are we trying to reduce? 

Reducing duplication in terms of code? Maybe auxilliary functions (although those can be made common). Register definitions though can just be done in header. This would be pretty easy to achieve just having the physical vs emulated driver side-by-side. In fact, it would be pretty easy to test an emulated peripheral 

If you mean duplication in terms of device tree, we have a fair bit of that already, but it's manageable, especially with inheritance (e.g. bosch,bmi160-i2c.yml, bosch,bmi160-i2c-emul.yml).

> ```
> &i2c0 {
>   dev@0x80 {
>     reg = <0x80>;
>     ...
>   };
> };
> ```
> 
> In this example, we're able to:
> 
> 1. Create the device struct/driver
> 2. Create the emulator
> 3. Automatically bind the emulator to the bus

All of those things are just general device tree and driver features already.

> 
> If we used a different compat string, I'm assuming you would want a separate node. In that case, you would need to duplicate the properties of the driver node. Also, if you have 2 nodes you can't put them both on the same bus since you can only have 1 node on the same address.

It can be, or it could be the same node name, or it could just be a chosen node (which is preferred). Duplication is something that we manage with DT already - even just using C-style `#includes` in DT can help to reduce that. 

I'm not sure if I follow w.r.t. having the same 2 nodes on the same bus at the same address, etc, because that shouldn't really be possible physically or logically. What I would do though, if testing a collection of emulated devices, is create a separate, mostly identical board definition, use emulated compat strings, and also use chosen properties.

I'm fairly certain that would achieve all of the same goals while reusing the usual DT machinery.

alevkoy commented:
@galak Are you still OK with this in its current form with `DEVICE_*` and `EMUL_*`?

mbolivar-nordic reviewed (approved):
Fingers crossed to get this in soon...

galak commented:
> @galak Are you still OK with this in its current form with `DEVICE_*` and `EMUL_*`?

Not sure I follow the question.
